src/pages/library/datasetBuilder/CohortEditor.ts

```diff
@@ -0,0 +1,442 @@
+        marginTop: wideMargin,
+        marginBottom: narrowMargin,
+        borderBottom: `1px solid ${colors.dark(0.35)}`,
+      },
+    },
+    [
+      div({ style: { margin: '5px', display: 'flex', alignItems: 'center' } }, [
+        h(
+          Link,
+          {
+            'aria-label': 'delete criteria',
+            onClick: () => {
+              deleteCriteria(criteria);
+            },
+          },
+          [icon('minus-circle-red', { size: 24, style: { color: colors.danger() } })]
+        ),
+        div({ style: { marginLeft: narrowMargin } }, [
+          Utils.cond(
+            [
+              'domainOption' in criteria,
+              () => {
+                const domainCriteria = criteria as DomainCriteria;
+                return h(Fragment, [strong([`${domainCriteria.domainOption.category}:`]), ` ${domainCriteria.name}`]);
+              },
+            ],
+            [
+              'listOption' in criteria,
+              () => {
+                const listCriteria = criteria as ProgramDataListCriteria;
+                return h(Fragment, [strong([`${criteria.name}:`]), ` ${listCriteria.value.name}`]);
+              },
+            ],
+            [
+              'rangeOption' in criteria,
+              () => {
+                const rangeCriteria = criteria as ProgramDataRangeCriteria;
+                return h(Fragment, [strong([`${criteria.name}:`]), ` ${rangeCriteria.low} - ${rangeCriteria.high}`]);
+              },
+            ],
+            [Utils.DEFAULT, () => div(['Unknown criteria type'])]
+          ),
+        ]),
+      ]),
+      `Count: ${criteria.count}`,
+    ]
+  );
+};
+
+export const createCriteriaViewComponent =
+  (deleteCriteria: (criteria: AnyCriteria) => void) => (criteria: AnyCriteria) =>
+    h(CriteriaView, { deleteCriteria, criteria, key: criteria.id });
+
+let criteriaCount = 1;
+const selectDomainCriteria = (domainType: DomainOption): DomainCriteria => {
+  // This needs to be replaced with a UI that lets users select the criteria they want from
+  // the list of concepts for this domain.
+  return {
+    domainOption: domainType,
+    name: domainType.values[0],
+    id: criteriaCount++,
+    // Need to call the API service to get the count for the criteria.
+    count: 100,
+  };
+};
+
+const createDefaultListCriteria = (listType: ProgramDataListOption): ProgramDataListCriteria => {
+  return {
+    listOption: listType,
+    name: listType.name,
+    id: criteriaCount++,
+    count: 100,
+    value: listType.values[0],
+  };
+};
+
+const createDefaultRangeCriteria = (rangeType: ProgramDataRangeOption): ProgramDataRangeCriteria => {
+  return {
+    rangeOption: rangeType,
+    name: rangeType.name,
+    id: criteriaCount++,
+    count: 100,
+    low: rangeType.min,
+    high: rangeType.max,
+  };
+};
+
+type CriteriaType = DomainOption | ProgramDataRangeOption | ProgramDataListOption;
+
+export function createCriteriaFromType(type: CriteriaType): AnyCriteria {
```

Comment:
The return type here seems deterministic on the type argument's type.  This can be expressed with function overloading on the type signatures.  Lookup ways to define function signature overloads in Typescript.  This will remove the need for some of the "as TypeX" casting I see elsewhere.

Reply:
Nice, I didn't know that function overloads could also change return types, and adding them does let me remove `as` from the test code. However I get an error on line 183 below:
```
        const criteria = createCriteriaFromType(x.value);
```
error:
```
TS2769: No overload matches this call.
  Overload 1 of 3, '(type: DomainOption): DomainCriteria', gave the following error.
    Argument of type 'CriteriaOption' is not assignable to parameter of type 'DomainOption'.
      Property 'category' is missing in type 'ProgramDataListOption' but required in type 'DomainOption'.
  Overload 2 of 3, '(type: ProgramDataListOption): ProgramDataListCriteria', gave the following error.
    Argument of type 'CriteriaOption' is not assignable to parameter of type 'ProgramDataListOption'.
      Type 'DomainOption' is missing the following properties from type 'ProgramDataListOption': dataType, name
  Overload 3 of 3, '(type: ProgramDataRangeOption): ProgramDataRangeCriteria', gave the following error.
    Argument of type 'CriteriaOption' is not assignable to parameter of type 'ProgramDataRangeOption'.
      Type 'DomainOption' is missing the following properties from type 'ProgramDataRangeOption': dataType, min, max, name
 ```

Reply:
I fixed this, it turns out I needed to add another overload for the original function signature too. So I now have:
```
export function createCriteriaFromType(type: DomainOption): DomainCriteria;
export function createCriteriaFromType(type: ProgramDataListOption): ProgramDataListCriteria;
export function createCriteriaFromType(type: ProgramDataRangeOption): ProgramDataRangeCriteria;
export function createCriteriaFromType(type: CriteriaOption): AnyCriteria;

export function createCriteriaFromType(type: CriteriaOption): AnyCriteria {
[...]
```